include/nbl/video/utilities/ngfx.h

```diff
@@ -0,0 +1,85 @@
+#ifndef _NBL_VIDEO_UTILITIES_NGFX_H_INCLUDED_
+#define _NBL_VIDEO_UTILITIES_NGFX_H_INCLUDED_
+
+// TODO: hopefully this is temporary
+#include "C:\Program Files\NVIDIA Corporation\Nsight Graphics 2024.1.0\SDKs\NsightGraphicsSDK\0.8.0\include\NGFX_Injection.h"
+
+namespace nbl::video
+{
+    struct SNGFXIntegration
+    {
+        bool useNGFX;
+        NGFX_Injection_InstallationInfo versionInfo;
+    };
+
+    inline bool injectNGFXToProcess(SNGFXIntegration& api)
+    {
+        uint32_t numInstallations = 0;
+        auto result = NGFX_Injection_EnumerateInstallations(&numInstallations, nullptr);
+        if (numInstallations == 0 || NGFX_INJECTION_RESULT_OK != result)
+        {
+            api.useNGFX = false;
+            return false;
+        }
+
+        std::vector<NGFX_Injection_InstallationInfo> installations(numInstallations);
+        result = NGFX_Injection_EnumerateInstallations(&numInstallations, installations.data());
+        if (numInstallations == 0 || NGFX_INJECTION_RESULT_OK != result)
+        {
+            api.useNGFX = false;
+            return false;
+        }
+
+        // get latest installation
+        api.versionInfo = installations.back();
+
+        uint32_t numActivities = 0;
+        result = NGFX_Injection_EnumerateActivities(&api.versionInfo, &numActivities, nullptr);
+        if (numActivities == 0 || NGFX_INJECTION_RESULT_OK != result)
+        {
+            api.useNGFX = false;
+            return false;
+        }
+
+        std::vector<NGFX_Injection_Activity> activities(numActivities);
+        result = NGFX_Injection_EnumerateActivities(&api.versionInfo, &numActivities, activities.data());
+        if (NGFX_INJECTION_RESULT_OK != result)
+        {
+            api.useNGFX = false;
+            return false;
+        }
+
+        const NGFX_Injection_Activity* pActivityToInject = nullptr;
+        for (const NGFX_Injection_Activity& activity : activities)
+        {
+            if (activity.type == NGFX_INJECTION_ACTIVITY_FRAME_DEBUGGER)    // only want frame debugger
+            {
+                pActivityToInject = &activity;
+                break;
+            }
+        }
+
+        if (!pActivityToInject) {
+            api.useNGFX = false;
+            return false;
+        }
+
+        result = NGFX_Injection_InjectToProcess(&api.versionInfo, pActivityToInject);
+        if (NGFX_INJECTION_RESULT_OK != result)
+        {
+            api.useNGFX = false;
+            return false;
+        }
```

Comment:
what happens if you are already running through nsight? what do these functions return?

can you detect if you've already been injected into?

Reply:
Seems to just crash at the moment because it's already injected. Though, could make it just fail quietly.

Reply:
Actually, I think this injection happens first. Nsight seems to only begin injection once vulkan/everything (not sure how much) is initialized, at which point it crashes the program.

Reply:
> Seems to just crash at the moment because it's already injected. Though, could make it just fail quietly.

I don't want to inject NGFX when EXE is not launched with NSight.

I want Nsight to do the injection, and us to just know if it happened, and not do it ourselves.

Reply:
> Actually, I think this injection happens first. Nsight seems to only begin injection once vulkan/everything (not sure how much) is initialized, at which point it crashes the program. Because it should just fail quietly at the moment.

This means that you need to possibly move the Renderdoc and Nsight injection from IAPIConnection constructor to the base class constructor/initialization finish

Idea: have a `init_finish` private method dealign with Renderdoc and Nsight which the derived class needs to call after all the setup is complete (end of most derived ctor).